engine/src/main/java/com/datatorrent/stram/webapp/OperatorDiscoverer.java

```diff
@@ -295,17 +297,42 @@ public void buildTypeGraph()
             JarFile jar = new JarFile(path);
             openJarFiles.put(path, jar);
             java.util.Enumeration<JarEntry> entriesEnum = jar.entries();
+            outer:
             while (entriesEnum.hasMoreElements()) {
-              java.util.jar.JarEntry jarEntry = entriesEnum.nextElement();
-              if (!jarEntry.isDirectory() && jarEntry.getName().endsWith("-javadoc.xml")) {
+              final java.util.jar.JarEntry jarEntry = entriesEnum.nextElement();
+              String entryName = jarEntry.getName();
+              if (jarEntry.isDirectory()) {
+                continue;
+              }
+              if (entryName.endsWith("-javadoc.xml")) {
                 try {
                   processJavadocXml(jar.getInputStream(jarEntry));
                   // break;
                 } catch (Exception ex) {
-                  LOG.warn("Cannot process javadoc {} : ", jarEntry.getName(), ex);
+                  LOG.warn("Cannot process javadoc {} : ", entryName, ex);
+                }
+              } else if (entryName.endsWith(".class")) {
+                TypeGraph.TypeGraphVertex newNode = typeGraph.addNode(jarEntry, jar);
+                // check if any visited resources belong to this type
+                for (Iterator<String> iter = resSet.iterator(); iter.hasNext(); ) {
+                  String entry = iter.next();
+                  if (entry.startsWith(entryName.substring(0, entryName.length() - 6))) {
+                    newNode.setHasRes(true);
+                    iter.remove();
+                  }
+                }
+              } else {
+                String className = entryName;
+                // check if this resource belongs to any visited type
+                while (className.contains("/")) {
```

Comment:
This nested while loop makes this somewhat confusing (since now we have to break to the outer indicator). It seems like it should be possible to refactor this to avoid having to do this.